[docs] Added docs for vertx pg client for Java (#13567)
yugabyte/yugabyte-db: docs/content/preview/develop/build-apps/java/ysql-vertx-pg-client.md

      <i class="icon-postgres" aria-hidden="true"></i>
      YSQL - Vertx PG Client
    </a>
  </li>
  <li >
    <a href="../ysql-jdbc-ssl/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      YSQL - JDBC SSL/TLS
    </a>
  </li>
  <li >
    <a href="../ysql-hibernate/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      YSQL - Hibernate
    </a>
  </li>
  <li >
    <a href="../ysql-sdyb/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      YSQL - Spring Data YugabyteDB
    </a>
  </li>
  <li >
    <a href="../ysql-spring-data/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      YSQL - Spring Data JPA
    </a>
  </li>
  <li>
    <a href="../ysql-ebean/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      YSQL - Ebean
    </a>
  </li>
  <li>
    <a href="../ycql/" class="nav-link">
      <i class="icon-cassandra" aria-hidden="true"></i>
      YCQL
    </a>
  </li>
  <li>
    <a href="../ycql-4.6/" class="nav-link">
      <i class="icon-cassandra" aria-hidden="true"></i>
      YCQL (4.6)
    </a>
  </li>
</ul>

> [Vertx PG client](https://vertx.io/docs/vertx-pg-client/java/) is the client for PostgreSQL with the simple APIs for communicating with the Database. It is reactive and non-blocking client for handling the database connections with single threaded API.

Reviewer: ```suggestion
[Vertx PG Client](https://vertx.io/docs/vertx-pg-client/java/) is the client for PostgreSQL with simple APIs to communicate with the database. It is a reactive and non-blocking client for handling the database connections with a single threaded API.
```

> As this client is for PostgreSQL and YugabyteDB is wire compatible with PostgreSQL, it is also supported by YugabyteDB with one [limitation](#limitation).

Reviewer: Because Vertx is a PostgreSQL client and YugabyteDB is PostgreSQL compatible, it is supported by YugabyteDB with a [limitation](#limitation)

## Prerequisites

This tutorial assumes that:

> - YugabyteDB is up and running. If you are new to YugabyteDB, you can download, install, and have YugabyteDB up and running within five minutes by following the steps in [Quick start](../../../../quick-start/).

Reviewer: ```suggestion
- YugabyteDB up and running. Download and install YugabyteDB by following the steps in [Quick start](../../../../quick-start/).
```

- Java Development Kit (JDK) 1.8, or later, is installed. 
- [Apache Maven](https://maven.apache.org/index.html) 3.3 or later, is installed.

## Create and configure the Java project

1. Create a project called "MySample".

    ```sh
    $ mvn archetype:generate \
        -DgroupId=com.yugabyte \
        -DartifactId=vertx-pg-example \
        -DarchetypeArtifactId=maven-archetype-quickstart \
        -DinteractiveMode=false

    $ cd vertx-pg-example
    ```

> 1. Add the following dependency for the vertx pg client within the `<dependencies>` element in the `pom.xml` file.

Reviewer: ```suggestion
1. Add the following dependency for the Vertx PG Client within the `<dependencies>` element in the `pom.xml` file.
```

    ```xml
    <dependency>
        <groupId>io.vertx</groupId>
        <artifactId>vertx-pg-client</artifactId>
        <version>4.3.2</version>
    </dependency>
    ```

1. Install the added dependency.

    ```sh
    $ mvn install
    ```

## Create the sample Java application

> 1. Copy the following Java code to a new file named `src/main/java/com/yugabyte/vertxPgExample.java`:

Reviewer: ```suggestion
1. Copy the following Java code to a new file `src/main/java/com/yugabyte/vertxPgExample.java`:
```

                            return preparedStatement.createStream(50, Tuple.tuple());
                        })
                        .onComplete(ar -> {
                            if (ar.succeeded()) {
                                RowStream<Row> stream = ar.result();
                                stream
                                        .exceptionHandler(promise::fail)
                                        .endHandler(promise::complete)
                                        .handler(row -> System.out.println(row.toJson())); // Printing each row as JSON
                            } else {
                                promise.fail(ar.cause());
                            }
                        });
                return promise.future().onComplete(v -> {
                    // close the connection
                    connection.close();
                });
            }).onComplete(ar -> {
                if (ar.succeeded()) {
                    System.out.println("Example ran successfully!");
                } else {
                    ar.cause().printStackTrace();
                }
            });

        }
    }
    ```

> 1. Run your new program.

Reviewer: ```suggestion
1. Run the program.
```

    ```sh
    $ mvn -q package exec:java -DskipTests -Dexec.mainClass=com.yugabyte.vertxPgExample
    ```

    You should see the following as the output:

    ```output
    {"id":1,"name":"Hello"}
    {"id":2,"name":"World"}
    {"id":3,"name":"Example"}
    {"id":4,"name":"Vertx"}
    {"id":5,"name":"Yugabyte"}
    Example ran successfully!
    ```
## Limitation

> [PubSub](https://vertx.io/docs/vertx-pg-client/java/#_pubsub) feature of Vertx PG client is currently not supported with YugabyteDB.

Reviewer: If there's a GitHub issue associated with this limitation, include it.

Reviewer 2: @priyanshi-yb You can include this issue: #1872

> [PubSub](https://vertx.io/docs/vertx-pg-client/java/#_pubsub) feature of Vertx PG client is currently not supported with YugabyteDB.

Reviewer: ```suggestion
[Pub/sub](https://vertx.io/docs/vertx-pg-client/java/#_pubsub) feature of Vertx PG client is currently not supported with YugabyteDB.
```

      YSQL - JDBC SSL/TLS
    </a>
  </li>
  <li >
    <a href="../ysql-hibernate/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      YSQL - Hibernate
    </a>
  </li>
  <li >
    <a href="../ysql-sdyb/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      YSQL - Spring Data YugabyteDB
    </a>
  </li>
  <li >
    <a href="../ysql-spring-data/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      YSQL - Spring Data JPA
    </a>
  </li>
  <li>
    <a href="../ysql-ebean/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      YSQL - Ebean
    </a>
  </li>
  <li>
    <a href="../ycql/" class="nav-link">
      <i class="icon-cassandra" aria-hidden="true"></i>
      YCQL
    </a>
  </li>
  <li>
    <a href="../ycql-4.6/" class="nav-link">
      <i class="icon-cassandra" aria-hidden="true"></i>
      YCQL (4.6)
    </a>
  </li>
</ul>

[Vertx PG Client](https://vertx.io/docs/vertx-pg-client/java/) is the client for PostgreSQL with simple APIs to communicate with the database. It is a reactive and non-blocking client for handling the database connections with a single threaded API.

> Because Vertx is a PostgreSQL client and YugabyteDB is PostgreSQL compatible, it is supported by YugabyteDB with a [limitation](#limitation).

Reviewer 3: `Because Vertx is a PostgreSQL client and YugabyteDB is PostgreSQL compatible, it is supported by YugabyteDB with a [limitation](#limitation).`

Suggestion:
`Since YugabyteDB is wire compatible with PostgreSQL, Vertx PG Client works with YugabyteDB as well.`

And skip mentioning the limitation here itself since users will know about it as they go through rest of the page, IMO.

Author: Okay, that makes sense.

                                        .endHandler(promise::complete)
                                        .handler(row -> System.out.println(row.toJson())); // Printing each row as JSON
                            } else {
                                promise.fail(ar.cause());
                            }
                        });
                return promise.future().onComplete(v -> {
                    // close the connection
                    connection.close();
                });
            }).onComplete(ar -> {
                if (ar.succeeded()) {
                    System.out.println("Example ran successfully!");
                } else {
                    ar.cause().printStackTrace();
                }
            });

        }
    }
    ```

1. Run the program.

    ```sh
    $ mvn -q package exec:java -DskipTests -Dexec.mainClass=com.yugabyte.vertxPgExample
    ```

    You should see the following as the output:

    ```output
    {"id":1,"name":"Hello"}
    {"id":2,"name":"World"}
    {"id":3,"name":"Example"}
    {"id":4,"name":"Vertx"}
    {"id":5,"name":"Yugabyte"}
    Example ran successfully!
    ```
## Limitation

> [PubSub](https://vertx.io/docs/vertx-pg-client/java/#_pubsub) feature of Vertx PG client is currently not supported with YugabyteDB. This limitation will be lifted when `LISTEN`/`NOTIFY` support is added to YugabyteDB. Tracking issue: [#1872](https://github.com/yugabyte/yugabyte-db/issues/1872).

Reviewer: ```suggestion
[Pub/sub](https://vertx.io/docs/vertx-pg-client/java/#_pubsub) feature of Vertx PG client is currently not supported with YugabyteDB. This limitation will be lifted when `LISTEN`/`NOTIFY` support is added to YugabyteDB. See [GitHub issue](https://github.com/yugabyte/yugabyte-db/issues/1872).
```
